metaphor/kafka/schema_parsers/protobuf_parser.py

```diff
@@ -0,0 +1,157 @@
+        file_path = self.proto_interface_dir / (schema_name + ".proto")
+        with open(file_path.as_posix(), "w") as file:
+            file.write(raw_schema)
+        return file_path
+
+    def _get_protobuf_python_object(self, file_path: Path, schema_name: str):
+        """
+        Compiles the protobuf schema to a Python file, then import the Python
+        file dynamically to determine the top level message object.
+
+        The top level message name is parsed internally. If there are more than
+        one possible top level message, we use the first message type that isn't
+        referenced by any other ones as our top level message.
+        """
+
+        # Instead of using grpc_tool.protoc.main, we directly invoke so that
+        # we don't need to specify a bunch of imports.
+        subprocess.run(
+            [
+                sys.executable,  # Use this or can't run pytest properly
+                "-m",
+                "grpc_tools.protoc",
+                f"--proto_path=generated={self.proto_interface_dir.as_posix()}",
+                f"--python_out={self.base_dir}",
+                file_path.as_posix(),
+            ]
+        )
+
+        sys.path.append(self.generated_dir.as_posix())
+        py_file = self.generated_dir / f"{schema_name.replace('-', '_')}_pb2.py"
+        module_name = py_file.stem
+        message = importlib.import_module(module_name)
+        top_level_message = self._find_top_level_message(message)
+        if not top_level_message:
+            raise ValueError("Cannot determine top level message")
+        class_ = getattr(message, top_level_message)
+        instance = class_()
+        return instance
+
+    def _find_top_level_message(self, message) -> Optional[str]:
+        # Phew that's a lot of reflection
+
+        file_desc: FileDescriptor = message.DESCRIPTOR
+
+        # Collect message types that are referenced in other messages
+        referenced_message_types = set()
+        for message in file_desc.message_types_by_name.values():
+            for field in message.fields:
+                if (
+                    field.type == 11
+                    and field.message_type is not None
+                    and str(field.message_type.full_name).startswith(file_desc.package)
+                ):
+                    referenced_message_types.add(field.message_type.full_name)
+
+        # Find the first message that isn't referenced in other messages
+        for message in file_desc.message_types_by_name.values():
+            if message.full_name not in referenced_message_types:
+                return message.name
+
+        return None  # If no top-level message is found
+
+    def _parse_protobuf_fields(self, fields, cur_path: str) -> List[SchemaField]:
+        schema_fields = []
+        for field in fields:
+            field_path = ".".join([cur_path, field.name])
+            if cur_path == "":
+                field_path = field_path[1:]
```

Comment:
Not sure if there is a simpler way to drop the leading `sep` if the first array element is empty.
```suggestion
            field_path = ".".join([cur_path, field.name]) if cur_path else field.name
```